### DISCUSSION TITLE
DAI 2024-2025 - Practical work 1 - Pedro Alves da Silva and Gonçalo Carvalheiro Heleno

### DISCUSSION THREAD

**Discussion by PedroAS7** [heig-vd-dai-course/heig-vd-dai-course#328] [Show and tell]
# Summary

We intend to create a command-line program that takes an image and a text message as input, hides the message inside the image and finally writes the file for the new image.

## Possible implementation

- Parse a bitmap image file by reading its header
- Verify that the image size can accommodate the message inside.
- Read bitmap's data (_pixel array_) and store it in a buffer
- Read the text message bit by bit and store it in a buffer
- Replace the 1 to 2 less significant bits (to decide how many) of each color for every pixel (or simply one of the colors for each pixel) with 1 to 2 bits from the text message (read sequentially).
- Output the stored buffer as a new image.

## Feasibility study

We've discussed about using [JPEG](https://en.wikipedia.org/wiki/JPEG#Encoding), [PNG](https://en.wikipedia.org/wiki/PNG#File_format), and [BMP](https://en.wikipedia.org/wiki/BMP_file_format) images to hide an image. However because both JPEG and PNG are compressed by default, are way more complex than BMP, and / or rely on a bunch of different sections (called _chunks_) that are "guarded" by a checksum, we've decided to go with the bitmap format.

The BMP file format is very simple and straight-forward. It is made of
- A mandatory header, containing a _magic byte_, file size, and _pixel array_ size and offset
- A second mandatory header (_DIB header_) which gives detailed information about the image (compression, has transparency, gamma, etc.)
- A few optional headers
- A _pixel array_ which contains the data that will hold the message we want to hide

The bitmap's pixel array is a simple list of bytes comprised of, in case a BMP file is encoded in the R8 G8 B8 format (8-bits for every color), 3 bytes, each of them containing a color. It does not follow the RGB order - it follows a BGR order (RGB but reversed). Furthermore, the first pixel in the pixel array is not the top-left pixel - it's the bottom-left pixel.

It is, therefore, simple enough to edit this file format to hide a message.

## Eventual improvements

- Provide an argument to set how many bits are to be stored in each pixel color byte
- Provide a way to perform the opposite operation and output a message from an encoded image
- Encrypt the message before storing it inside the image by using an encryption key passed as an argument

## Repository & tag

Practical work repository is available at https://github.com/lentidas/DAI-2425-PW1
Release tag is https://github.com/lentidas/DAI-2425-PW1/releases/tag/v1.0.0

**Comment by GeraudSilvestri** [comment]
It is good for me. Be wary that the time is limited, so focus on your main functionality and when that is done and clean, add the improvements

**Comment by lentidas** [comment]
Here it is the tag of our repo that we submit for evaluation-> https://github.com/lentidas/DAI-2425-PW1/releases/tag/v1.0.0
